### DISCUSSION TITLE
Conda environment using user site-packages: `ImportError: DLL load failed: The specified module could not be found.`

### DISCUSSION THREAD

**Discussion by gracecm** [talmolab/sleap#829] [Help!] [answered]
# Bug description
Hello! I am having trouble opening my install of sleap 1.2.4, I think it is failing to load cv2. 

## Expected behaviour

Entered sleap-label in newly installed environment, expected it to open the gui as usual.

## Actual behaviour
Gui did not open, seemed to be stuck on loading cv2 module.

## Your personal set up

- OS:

[Windows 10]
And I updated my GPU driver
- Version(s):
[SLEAP v 1.2.4, python 3.7.12, Tensorflow 2.6.3]

  - [ ] [Conda from package](https://sleap.ai/installation.html#conda-package)
I 
<details><summary>Environment packages</summary>

```
# paste output of `pip freeze` or `conda list` here
# Name                    Version                   Build  Channel
absl-py                   0.15.0                   pypi_0    pypi
astunparse                1.6.3                    pypi_0    pypi
attrs                     21.2.0                   pypi_0    pypi
backports-zoneinfo        0.2.1                    pypi_0    pypi
ca-certificates           2022.6.15            h5b45459_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cachetools                4.2.4                    pypi_0    pypi
cattrs                    1.1.1                    pypi_0    pypi
certifi                   2021.10.8                pypi_0    pypi
charset-normalizer        2.0.12                   pypi_0    pypi
clang                     5.0                      pypi_0    pypi
colorama                  0.4.5                    pypi_0    pypi
commonmark                0.9.1                    pypi_0    pypi
cuda-nvcc                 11.3.58              hb8d16a4_0    nvidia
cudatoolkit               11.3.1              h280eb24_10    conda-forge
cudnn                     8.2.1.32             h754d62a_0    conda-forge
cycler                    0.11.0                   pypi_0    pypi
efficientnet              1.0.0                    pypi_0    pypi
ffmpeg                    4.3.1                ha925a31_0    conda-forge
flatbuffers               1.12                     pypi_0    pypi
fonttools                 4.33.3                   pypi_0    pypi
freetype                  2.10.4               h546665d_1    conda-forge
gast                      0.4.0                    pypi_0    pypi
geos                      3.9.1                h39d44d4_2    conda-forge
google-auth               1.35.0                   pypi_0    pypi
google-auth-oauthlib      0.4.6                    pypi_0    pypi
google-pasta              0.2.0                    pypi_0    pypi
grpcio                    1.44.0                   pypi_0    pypi
h5py                      3.1.0           nompi_py37h19fda09_100    conda-forge
hdf5                      1.10.6          nompi_he0bbb20_101    conda-forge
idna                      3.3                      pypi_0    pypi
image-classifiers         1.0.0                    pypi_0    pypi
imageio                   2.15.0                   pypi_0    pypi
imgaug                    0.4.0                    pypi_0    pypi
imgstore                  0.2.9                    pypi_0    pypi
importlib-metadata        4.11.1                   pypi_0    pypi
intel-openmp              2022.1.0          h57928b3_3787    conda-forge
joblib                    1.1.0                    pypi_0    pypi
jpeg                      9e                   h8ffe710_2    conda-forge
jsmin                     3.0.1                    pypi_0    pypi
jsonpickle                1.2                      pypi_0    pypi
keras                     2.6.0                    pypi_0    pypi
keras-applications        1.0.8                    pypi_0    pypi
keras-preprocessing       1.1.2                    pypi_0    pypi
kiwisolver                1.4.3                    pypi_0    pypi
lcms2                     2.12                 h2a16943_0    conda-forge
lerc                      3.0                  h0e60522_0    conda-forge
libblas                   3.9.0              15_win64_mkl    conda-forge
libcblas                  3.9.0              15_win64_mkl    conda-forge
libdeflate                1.10                 h8ffe710_0    conda-forge
liblapack                 3.9.0              15_win64_mkl    conda-forge
libpng                    1.6.37               h1d00b33_2    conda-forge
libtiff                   4.3.0                hc4061b1_4    conda-forge
libzlib                   1.2.12               h8ffe710_1    conda-forge
lz4-c                     1.9.3                h8ffe710_1    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
markdown                  3.3.6                    pypi_0    pypi
matplotlib                3.5.2                    pypi_0    pypi
mkl                       2022.1.0           h6a75c08_874    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
networkx                  2.6.3                    pypi_0    pypi
numpy                     1.19.5           py37h4c2b6ed_3    conda-forge
oauthlib                  3.2.0                    pypi_0    pypi
olefile                   0.46               pyh9f0ad1d_1    conda-forge
opencv-python             4.5.5.62                 pypi_0    pypi
opencv-python-headless    4.5.5.62                 pypi_0    pypi
openjpeg                  2.4.0                hb211442_1    conda-forge
openssl                   3.0.5                h8ffe710_0    conda-forge
opt-einsum                3.3.0                    pypi_0    pypi
packaging                 21.3                     pypi_0    pypi
pandas                    1.3.5            py37h9386db6_0    conda-forge
pillow                    8.4.0            py37hd7d9ad0_0    conda-forge
pip                       22.0.3             pyhd8ed1ab_0    conda-forge
protobuf                  3.19.4                   pypi_0    pypi
psutil                    5.9.1                    pypi_0    pypi
pyasn1                    0.4.8                    pypi_0    pypi
pyasn1-modules            0.2.8                    pypi_0    pypi
pygments                  2.12.0                   pypi_0    pypi
pykalman                  0.9.5                    pypi_0    pypi
pyparsing                 3.0.7                    pypi_0    pypi
pyreadline                2.1             py37h03978a9_1006    conda-forge
pyside2                   5.14.1                   pypi_0    pypi
python                    3.7.12          h900ac77_100_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-rapidjson          1.6                      pypi_0    pypi
python_abi                3.7                     2_cp37m    conda-forge
pytz                      2022.1             pyhd8ed1ab_0    conda-forge
pytz-deprecation-shim     0.1.0.post0              pypi_0    pypi
pywavelets                1.3.0                    pypi_0    pypi
pyzmq                     23.2.0                   pypi_0    pypi
qimage2ndarray            1.8.3                    pypi_0    pypi
requests                  2.27.1                   pypi_0    pypi
requests-oauthlib         1.3.1                    pypi_0    pypi
rich                      10.16.1                  pypi_0    pypi
scikit-image              0.19.3                   pypi_0    pypi
scikit-learn              1.0.2                    pypi_0    pypi
scikit-video              1.1.11                   pypi_0    pypi
scipy                     1.7.3            py37hb6553fb_0    conda-forge
seaborn                   0.11.2                   pypi_0    pypi
segmentation-models       1.0.1                    pypi_0    pypi
setuptools                59.8.0           py37h03978a9_1    conda-forge
setuptools-scm            6.4.2                    pypi_0    pypi
shapely                   1.7.1            py37hc520ffa_5    conda-forge
shiboken2                 5.14.1                   pypi_0    pypi
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sleap                     1.2.4                    pypi_0    pypi
sqlite                    3.39.0               h8ffe710_0    conda-forge
tbb                       2021.5.0             h2d74725_1    conda-forge
tensorboard               2.6.0                    pypi_0    pypi
tensorboard-data-server   0.6.1                    pypi_0    pypi
tensorboard-plugin-wit    1.8.1                    pypi_0    pypi
tensorflow                2.6.3                    pypi_0    pypi
tensorflow-estimator      2.6.0                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
threadpoolctl             3.1.0                    pypi_0    pypi
tifffile                  2021.11.2                pypi_0    pypi
tk                        8.6.12               h8ffe710_0    conda-forge
tomli                     2.0.1                    pypi_0    pypi
typing-extensions         3.10.0.2                 pypi_0    pypi
tzdata                    2022.1                   pypi_0    pypi
tzlocal                   4.2                      pypi_0    pypi
ucrt                      10.0.20348.0         h57928b3_0    conda-forge
urllib3                   1.26.8                   pypi_0    pypi
vc                        14.2                 hb210afc_6    conda-forge
vs2015_runtime            14.29.30037          h902a5da_6    conda-forge
werkzeug                  2.0.3                    pypi_0    pypi
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
wrapt                     1.12.1                   pypi_0    pypi
xz                        5.2.5                h62dcd97_1    conda-forge
zipp                      3.7.0                    pypi_0    pypi
zlib                      1.2.12               h8ffe710_1    conda-forge
zstd                      1.5.2                h6255e5f_2    conda-forge
```

</details>

<details><summary>Logs</summary>

```
# paste relevant logs here, if any
Traceback (most recent call last):
  File "C:\Anaconda3\envs\sleap_id\Scripts\sleap-label-script.py", line 33, in <module>
    sys.exit(load_entry_point('sleap==1.2.4', 'console_scripts', 'sleap-label')())
  File "C:\Anaconda3\envs\sleap_id\Scripts\sleap-label-script.py", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "C:\Anaconda3\envs\sleap_id\lib\site-packages\importlib_metadata\__init__.py", line 194, in load
    module = import_module(match.group('module'))
  File "C:\Anaconda3\envs\sleap_id\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Anaconda3\envs\sleap_id\lib\site-packages\sleap\__init__.py", line 10, in <module>
    from sleap.io.dataset import Labels, load_file
  File "C:\Anaconda3\envs\sleap_id\lib\site-packages\sleap\io\dataset.py", line 69, in <module>
    from sleap.instance import (
  File "C:\Anaconda3\envs\sleap_id\lib\site-packages\sleap\instance.py", line 32, in <module>
    from sleap.io.video import Video
  File "C:\Anaconda3\envs\sleap_id\lib\site-packages\sleap\io\video.py", line 7, in <module>
    import cv2
  File "C:\Users\ShaevitzLab\AppData\Roaming\Python\Python37\site-packages\cv2\__init__.py", line 3, in <module>
    from .cv2 import *
ImportError: DLL load failed: The specified module could not be found.
```
</details>

## How to reproduce

1. Fresh install of SLEAP using the conda package
2. attempt command sleap-label

**Comment by roomrys** [comment, marked as answer]
Hi @gracecm,

**TLDR**
It seems [the workaround](https://stackoverflow.com/a/59508454/13281260) is to run the command [`set PYTHONNOUSERSITE=True`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONNOUSERSITE) (`set` for Windows, `export` for Linux) before running `conda activate <name of environment>` which will prevent the conda environment from accessing user site-package.

---

By looking at the path `"C:\Users\ShaevitzLab\AppData\Roaming\Python\Python37\site-packages\cv2"`, it seems that python37 is attempting to import your global install of `cv2`. We want python to use the conda environment's pip install of `cv2` which would be located at `"C:\Anaconda3\envs\sleap_id\lib\site-packages\cv2"`. From the output of `conda list`, we know that you have `opencv-python` and  `opencv-python-headless` installed in the environment, we just need to get python37 to use this installation instead of the global installation.

This issue arises because you have the same version of python installed globally as we currently use in SLEAP (python37). Per [PEP370](https://peps.python.org/pep-0370/), to be considered a "standard python distribution", conda needs to allow same versions of python to access globally installed python packages - there is a whole [list of issues](https://github.com/conda/conda/issues/8770#issuecomment-501310731) revolving around this on conda's github, but it's a PEP standard so not a whole lot to do other than manually disable access to user site-packages.

---

Let us know how that works.

Thanks,
Liezl

**Comment by gracecm** [comment]
Wonderful, this fixed it, thank you so much!
